## Data Cleaning for general purpose analyzing

The following site was used to perform data cleaning:
https://opendata.cityofnewyork.us/data/#datasetscategory

In [110]:
import os  
#Importing os module in python provides a way of using system depended functionality.

import numpy as np 
#Numpy is the core library for scientific computing in Python. np is an alias pointing to numpy

import pandas as pd 
#Pandas is widely used Python library in data science mainly used for data munging.

import seaborn as sns 
#Seaborn is a Python visualization library based on matplotlib providing a high-level interface for drawing attractive statistical graphics.

In [111]:
fp = 'C:/Vaibhavi/IDS/SAT_Results.csv'
results = pd.read_csv(fp)
results.head(20)
#head(): displays the top rows of the data, in this case top 20

DBN                                        SCHOOL NAME  \
0   01M292      HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1   01M448                UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2   01M450                         EAST SIDE COMMUNITY SCHOOL   
3   01M458                          FORSYTH SATELLITE ACADEMY   
4   01M509                            MARTA VALLE HIGH SCHOOL   
..     ...                                                ...   
15  02M303                         FACING HISTORY SCHOOL, THE   
16  02M305  URBAN ASSEMBLY ACADEMY OF GOVERNMENT AND LAW, THE   
17  02M308                       LOWER MANHATTAN ARTS ACADEMY   
18  02M313  JAMES BALDWIN SCHOOL, THE: A SCHOOL FOR EXPEDI...   
19  02M316  URBAN ASSEMBLY SCHOOL OF BUSINESS FOR YOUNG WO...   

   Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                      29                             355                 404   
1                      91                             383                 423   
2                      70                             377                 402   
3                       7                             414                 401   
4                      44                             390                 433   
..                    ...                             ...                 ...   
15                     76                             353                 358   
16                     50                             375                 388   
17                     40                             403                 392   
18                     69                             408                 390   
19                     42                             373                 370   

   SAT Writing Avg. Score  
0                     363  
1                     366  
2                     370  
3                     359  
4                     384  
..                    ...  
15                    340  
16                    385  
17                    405  
18                    390  
19                    384  

[20 rows x 6 columns]

In [112]:
column_names = {'DBN': 'DBN',  'SCHOOL NAME':'School Name',
                 'Num of SAT Test Takers': 'No of SAT Test Takers', 'SAT Critical Reading Avg. Score':'SAT Critical Reading Avg Score',
                'SAT Math Avg. Score':'SAT Math Avg Score',
                'SAT Writing Avg. Score':'SAT Writing Avg Score'}   

results = (results.rename(columns=column_names)
    .dropna(thresh=4)
    [['DBN', 'School Name', 'No of SAT Test Takers', 'SAT Critical Reading Avg Score', 'SAT Math Avg Score','SAT Writing Avg Score']]
        .set_index('DBN', append=True)
     .rename_axis(["Sr No", "DBN"])
    .sort_index())

#rename() : function takes as argument the dict of new column names
#dropna() : Returns object with labels on given axis omitted where alternately any or all of the data are missing
#assign() : function helps to change the date format using to_datetime function of pandas
#set_index() :Sets the DataFrame index (row labels) using one or more existing columns and by default yields a new object.
#rename_axis() :Alters index and / or columns using input function or functions.A scalar or list-like for mapper will alter the Index.name or MultiIndex.names attribute.
#sort_index() : Sorts object by labels (along an axis)
 
results.head() 

School Name  \
Sr No DBN                                                     
0     01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1     01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2     01M450                     EAST SIDE COMMUNITY SCHOOL   
3     01M458                      FORSYTH SATELLITE ACADEMY   
4     01M509                        MARTA VALLE HIGH SCHOOL   

             No of SAT Test Takers SAT Critical Reading Avg Score  \
Sr No DBN                                                           
0     01M292                    29                            355   
1     01M448                    91                            383   
2     01M450                    70                            377   
3     01M458                     7                            414   
4     01M509                    44                            390   

             SAT Math Avg Score SAT Writing Avg Score  
Sr No DBN                                              
0     01M292                404                   363  
1     01M448                423                   366  
2     01M450                402                   370  
3     01M458                401                   359  
4     01M509                433                   384

In [113]:
tidy = pd.melt(results.reset_index(),
               id_vars=['Sr No', 'DBN'], value_vars=['No of SAT Test Takers','SAT Math Avg Score','SAT Writing Avg Score'],
               value_name='sat')

#pd.melt : “Unpivots” a DataFrame from wide format to long format, optionally leaving identifier variables set.

tidy.head(100)

Sr No     DBN               variable sat
0       0  01M292  No of SAT Test Takers  29
1       1  01M448  No of SAT Test Takers  91
2       2  01M450  No of SAT Test Takers  70
3       3  01M458  No of SAT Test Takers   7
4       4  01M509  No of SAT Test Takers  44
..    ...     ...                    ...  ..
95     95  04M610  No of SAT Test Takers  70
96     96  04M635  No of SAT Test Takers  48
97     97  04M680  No of SAT Test Takers  31
98     98  05M285  No of SAT Test Takers  29
99     99  05M304  No of SAT Test Takers  54

[100 rows x 4 columns]

In [114]:
mask = results['No of SAT Test Takers'].map(lambda x: x != 's')
results = results[mask]

results['SAT Math Avg Score'] = results['SAT Math Avg Score'].astype(int)
results['SAT Critical Reading Avg Score'] = results['SAT Critical Reading Avg Score'].astype(int)
results['SAT Writing Avg Score'] = results['SAT Writing Avg Score'].astype(int)

results.head(600)


School Name  \
Sr No DBN                                                         
0     01M292      HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1     01M448                UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2     01M450                         EAST SIDE COMMUNITY SCHOOL   
3     01M458                          FORSYTH SATELLITE ACADEMY   
4     01M509                            MARTA VALLE HIGH SCHOOL   
...                                                         ...   
466   32K556  BUSHWICK LEADERS HIGH SCHOOL FOR ACADEMIC EXCE...   
467   32K564                     BUSHWICK COMMUNITY HIGH SCHOOL   
471   75Q811                                          P.S. Q811   
476   79Q950                                GED PLUS s CITYWIDE   
477   79X490                                    PHOENIX ACADEMY   

             No of SAT Test Takers  SAT Critical Reading Avg Score  \
Sr No DBN                                                            
0     01M292                    29                             355   
1     01M448                    91                             383   
2     01M450                    70                             377   
3     01M458                     7                             414   
4     01M509                    44                             390   
...                            ...                             ...   
466   32K556                    23                             347   
467   32K564                    24                             359   
471   75Q811                    32                             429   
476   79Q950                     8                             496   
477   79X490                     9                             367   

              SAT Math Avg Score  SAT Writing Avg Score  
Sr No DBN                                                
0     01M292                 404                    363  
1     01M448                 423                    366  
2     01M450                 402                    370  
3     01M458                 401                    359  
4     01M509                 433                    384  
...                          ...                    ...  
466   32K556                 358                    350  
467   32K564                 317                    358  
471   75Q811                 444                    433  
476   79Q950                 400                    426  
477   79X490                 370                    360  

[421 rows x 5 columns]

In [115]:
#Data munging – recap of the need
results.apply(lambda x: sum(x.isnull()),axis=0)  

#It shows there are no missing values

School Name                       0
No of SAT Test Takers             0
SAT Critical Reading Avg Score    0
SAT Math Avg Score                0
SAT Writing Avg Score             0
dtype: int64

In [116]:
results.describe()
#Get summary of Numerical Variables.

SAT Critical Reading Avg Score  SAT Math Avg Score  \
count                      421.000000          421.000000   
mean                       400.850356          413.368171   
std                         56.802783           64.684655   
min                        279.000000          312.000000   
25%                        368.000000          371.000000   
50%                        391.000000          395.000000   
75%                        416.000000          437.000000   
max                        679.000000          735.000000   

       SAT Writing Avg Score  
count             421.000000  
mean              393.985748  
std                58.635109  
min               286.000000  
25%               360.000000  
50%               381.000000  
75%               411.000000  
max               682.000000

Here are a few inferences, you can draw by looking at the output of describe() function:

1. School Names count is 421 which means all the non numeric School name entries are present in the data.
2. No of SAT Test Takers has 421 entries means all the entries are present.
3. The mean, min , std and max seems to be in line with expectation for all the values.

Please note that we can get an idea of a possible skew in the data by comparing the mean to the median, i.e. the 50% figure.

For the non-numerical values (e.g. School Name), we can look at frequency distribution to understand whether they make sense or not. The frequency table can be printed by following command:

In [40]:
results['School Name'].value_counts()

LIFE ACADEMY HIGH SCHOOL FOR FILM AND MUSIC                 1
PELHAM PREPARATORY ACADEMY                                  1
BROOKLYN COLLEGE ACADEMY                                    1
ACADEMY FOR CONSERVATION AND THE ENVIRONMENT                1
INTERNATIONAL HIGH SCHOOL AT LAGUARDIA COMMUNITY COLLEGE    1
                                                           ..
HIGH SCHOOL FOR HEALTH CAREERS AND SCIENCES                 1
ESSEX STREET ACADEMY                                        1
RACHEL CARSON HIGH SCHOOL FOR COASTAL STUDIES               1
ACADEMY OF URBAN PLANNING                                   1
ACADEMY FOR LANGUAGE AND TECHNOLOGY                         1
Name: School Name, Length: 478, dtype: int64

In [117]:
#Checking the frequency of numeric data

results['No of SAT Test Takers'].value_counts()

54     10
48      8
9       8
72      8
61      7
       ..
531     1
93      1
191     1
395     1
392     1
Name: No of SAT Test Takers, Length: 174, dtype: int64

In [102]:
results['SAT Critical Reading Avg Score'].value_counts()

384    8
398    8
367    8
370    8
373    7
      ..
341    1
339    1
337    1
336    1
513    1
Name: SAT Critical Reading Avg Score, Length: 163, dtype: int64

In [103]:
results['SAT Math Avg Score'].value_counts()

385    9
381    8
371    8
391    8
364    8
      ..
425    1
429    1
443    1
444    1
512    1
Name: SAT Math Avg Score, Length: 172, dtype: int64

In [104]:
results['SAT Writing Avg Score'].value_counts()

368    9
370    9
359    8
394    8
411    7
      ..
422    1
427    1
428    1
430    1
512    1
Name: SAT Writing Avg Score, Length: 162, dtype: int64

## References :

https://www.analyticsvidhya.com/blog/2016/01/complete-tutorial-learn-data-science-python-scratch-2/

https://data.cityofnewyork.us/Education/SAT-Results/f9bf-2cp4

https://opendata.cityofnewyork.us/data/#datasetscategory